Prompt to reconnect when a Jetpack disconnect is detected

mikkamp opened the pull request:
### Changes proposed in this Pull Request:

Project details: pcTzPl-oz-p2
This PR covers the feature branch for all the PR's:
- #1283 
- #1301 
- #1374
- #1385
- #1393
- #1414
- #1418

### Todo:
- [x] The notification action redirects to the setup page (once the UI part is implemented, this will need to be the reconnect page)
- [x] UI implementation

Closes #775

mikkamp committed:
Add Jetpack connected option

mikkamp committed:
Handle disconnected errors

Send custom exception when a disconnect has been detected. We still log the
original exception.

mikkamp committed:
Make controllers to use response_from_exception

This is used to return a standardized response regardless if it is a
generic exception or if it contains response data.
This is only done for the controllers who send requests to the WCS,
these can get a AccountReconnect exception thrown.

mikkamp committed:
Avoid warning if email is not inlcuded in status

mikkamp committed:
Unit test for the Jetpack account controller

mikkamp committed:
Unit test for the Google account controller

mikkamp committed:
Merge pull request #1283 from woocommerce/jetpack-disconnect-backend

Handle a disconnected Jetpack account in the backend

mikkamp committed:
Helper function to check if Jetpack is connected

mikkamp committed:
Remove unused import

mikkamp committed:
Delete function to allow a note to clear itself up

mikkamp committed:
Add WordPress reconnect notification

mikkamp committed:
Trigger notification on jetpack connection change

mikkamp committed:
Unit test for notification should add conditions

mikkamp committed:
Shorten notification text

mikkamp committed:
Reset Jetpack connected on disconnect

mikkamp committed:
Merge pull request #1301 from woocommerce/add/reconnect-jetpack-notification

Reconnect WordPress notification

mikkamp committed:
Merge develop into feature/jetpack-disconnect

eason9487 committed:
Move GoogleAdsAccountCard to shared component folder

eason9487 committed:
Internalize WPCOM appearance to AccountCard

eason9487 committed:
Move WPComAccountCard to shared component folder

eason9487 committed:
Move GoogleMCAccountCard to shared component folder

eason9487 committed:
Rename DisconnectAccounts to LinkedAccounts and swap its position with ContactInformation

eason9487 committed:
Change the linked accounts section on settings page to use Card UI to present each account

eason9487 committed:
Remove an unused file

eason9487 committed:
Update an e2e test case

eason9487 committed:
Rename reconnect-accounts to reconnect-google-account

eason9487 committed:
Change the URL subpath of reconnecting google account

eason9487 committed:
Add the `next` parameter to GET `jetpack/connect` API for indicating the redirect URI when back from Jetpack authorization

eason9487 committed:
Forward rest props of AccountCard to Section.Card, and adjust its style

eason9487 committed:
Make ConnectWPComAccountCard call `jetpack/connect` API with the `next` parameter.

eason9487 committed:
Add a new component for reconnecting WPCOM account

eason9487 committed:
Add the re-connecting flow for the disconnected WPCOM account

eason9487 committed:
Rename exposed components to make their internal names are as the same as exporting names.
- google-ads-account-card/connected.js -> google-ads-account-card/connected-google-ads-account-card.js
- google-mc-account-card/connected-card.js -> google-mc-account-card/connected-google-mc-account-card.js

Address https://github.com/woocommerce/google-listings-and-ads/pull/1374#discussion_r839466722

eason9487 committed:
Adjust the layout width on the Settings page to align with the visual design.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1374#issuecomment-1084447261

eason9487 committed:
Fix grammar problems in JSDoc.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1374#discussion_r839621525

eason9487 committed:
Adjust layouts of AccountCard and related components to make them much closer to the visual design.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1374#issuecomment-1084629176

eason9487 committed:
Update a description in the tracking README.md

Address https://github.com/woocommerce/google-listings-and-ads/pull/1385#discussion_r839588406

Co-authored-by: Miguel Pérez Pellicer <5908855+puntope@users.noreply.github.com>

eason9487 committed:
Rename the `next` parameters of Google and JetPack connect APIs to `next_page_name`.

Address
- https://github.com/woocommerce/google-listings-and-ads/pull/1385#discussion_r839567333
- https://github.com/woocommerce/google-listings-and-ads/pull/1385#discussion_r839589560
- https://github.com/woocommerce/google-listings-and-ads/pull/1385#discussion_r839596788

eason9487 committed:
Combine two functions that get the re-connecting account URL into one.

Address
- https://github.com/woocommerce/google-listings-and-ads/pull/1385#discussion_r839569258
- https://github.com/woocommerce/google-listings-and-ads/pull/1385#discussion_r839576276
- https://github.com/woocommerce/google-listings-and-ads/pull/1385#discussion_r839580484

eason9487 committed:
Fix PHP linting warnings

eason9487 committed:
Merge pull request #1374 from woocommerce/tweak/775-adjust-settings-page

Re-connecting WPCOM Account - Adjust the linked accounts section in settings

eason9487 committed:
Merge pull request #1385 from woocommerce/add/775-reconnect-wpcom-account

Re-connecting WPCOM Account - Add the re-connecting flow

eason9487 committed:
Merge branch 'develop' into feature/jetpack-disconnect

Resolved conflicts:
- git mv js/src/setup-ads/ads-stepper/setup-accounts/google-ads-account-card/connect-ads/ads-account-select-control.js js/src/components/google-ads-account-card/connect-ads/
- js/src/constants.js
- js/src/setup-ads/ads-stepper/setup-accounts/index.js
- src/API/Site/Controllers/Ads/CampaignController.php

eason9487 committed:
Fix the failed test in Google/AccountControllerTest.

It was missed when adding cabe35a5033fec920e3d91bcd96c011569f9cc1a

eason9487 committed:
Fix warnings in the test of Jetpack/AccountControllerTest.

After merfing develop branch, there is another Proxy class importing that needs to be renamed.
This commit followed the changes in e36f7ade791fbc8ba4452370e38168ba7bec7253 to do so.

eason9487 committed:
Merge pull request #1393 from woocommerce/fix/jetpack-disconnect-php-unit-tests

Fix PHP two unit tests for the shared branch of re-connecting WPCOM account

eason9487 commented:
> * The notification action redirects to the setup page (once the UI part is implemented, this will need to be the reconnect page)

The settings page can also detect 401 and then redirect to the reconnect page, so it's okay to let notification action redirect to the settings page. Or if we want to go directly to the reconnect page, its path is `admin.php?page=wc-admin&path=/google/settings&subpath=/reconnect-wpcom-account`

mikkamp committed:
Redirect directly to the reconnect page

mikkamp committed:
Confirm we have a valid token

mikkamp committed:
Add tests for invalid token

eason9487 committed:
Add a lock function for locking the available GLA link during reconnecting process

eason9487 committed:
Apply the lock and unlock functions to reconnecting processes

eason9487 committed:
Don't display the classic navigation tabs during the reconnecting process

mikkamp committed:
Supress errors when checking valid token

mikkamp committed:
Merge pull request #1414 from woocommerce/fix/1348-redirect-notification-to-reconnect

Redirect notification to reconnect page

eason9487 committed:
Change the export syntax for functions in the reconnectionLock.js

eason9487 committed:
Select GLA menu links by the href attribute instead of their positions.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1418#discussion_r847378470

eason9487 committed:
Merge branch 'develop' into feature/jetpack-disconnect

Edits:
- Add a tracking property `context` to JSDoc and run `npm run doc:tracking` to update the tracking doc.
   - js/src/components/wpcom-account-card/connect-wpcom-account-card.js
   - src/Tracking/README.md
- Make new tests to be PHPUnit9 compatible
   - tests/Unit/API/Site/Controllers/Google/AccountControllerTest.php
   - tests/Unit/API/Site/Controllers/Jetpack/AccountControllerTest.php
   - tests/Unit/Notes/ReconnectWordPressTest.php

Conflicts:
- js/src/components/google-mc-account-card/connected-google-mc-account-card.js
- js/src/settings/linked-accounts.js
- src/Tracking/README.md

eason9487 committed:
Merge pull request #1418 from woocommerce/feature/775-disable-menu-links

Re-connecting WPCOM Account - Disable menu links during the re-connecting process

eason9487 commented:
- Resolved the conflicts with `develop` - 0c8cb7f17696fe28a68b8a7742ddae3c19c37fae
- The frontend parts have been completed.
- Did a round of overall testing with the dfc5e80d314cf31686ef5939ab677a55f0ea64b6 revision of `feature/jetpack-disconnect`. Both WPCOM and Google account re-connecting processing work well.

mikkamp commented:
I did a round of testing including the UI when I implemented #1414 so I think this is good to go. As far as I know we don't have any blockers so this can be merged into develop to be included in the next release.
